pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -0,0 +1,200 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// TRIE data structure inspired by https://github.com/dghubble/trie
+/*
+	This differs from the original trie.
+   	This has been modified to detect partial matches as well.
+   	For eg.
+		If we add "ABCD" to this trie, and try to check if "ABCDEF" is in the trie,
+		it will return true because that's how fingerprint matching works in current implementation.
+*/
+
+package tracker // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/tracker"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/checkpoint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fileset"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/reader"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+)
+
+var errTooManyActiveFiles = errors.New("number of actively read files exceeds max_concurrent_files")
+
+type Tracker struct {
+	*zap.SugaredLogger
+
+	ReaderFactory reader.Factory
```

Comment:
I think that putting the reader factory within the tracker is giving it more responsibility than necessary. The primary concern of the tracker should be understanding uniqueness and state of files. Therefore, I think creating the readers should happen outside of the tracker.